File: firebase-storage/src/main/java/com/github/b3er/rxfirebase/storage/RxFirebaseStorage.java
package com.github.b3er.rxfirebase.storage;

import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import com.google.firebase.storage.FileDownloadTask;
import com.google.firebase.storage.StorageMetadata;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.StreamDownloadTask;
import com.google.firebase.storage.UploadTask;
import io.reactivex.Completable;
import io.reactivex.Single;
import java.io.File;
import java.io.InputStream;

public final class RxFirebaseStorage {

  private RxFirebaseStorage() {
    throw new AssertionError("No instances");
  }

  /**
   * @see StorageReference#getBytes(long)
   */
  public static Single<byte[]> getBytes(@NonNull StorageReference ref,
      final long maxDownloadSizeBytes) {
    return Single.create(new GetBytesOnSubscribe(ref, maxDownloadSizeBytes));
  }

  /**
   * @see StorageReference#delete()
   */
  public static Completable delete(@NonNull StorageReference ref) {
    return Completable.create(new DeleteOnSubscribe(ref));
  }

  /**
   * @see StorageReference#getFile(Uri)
   */
  public static Single<FileDownloadTask.TaskSnapshot> getFile(@NonNull StorageReference ref,
      @NonNull Uri uri) {
    return Single.create(new GetFileToUriOnSubscribe(ref, uri));
  }

  /**
   * @see StorageReference#getFile(File)
   */
  public static Single<FileDownloadTask.TaskSnapshot> getFile(@NonNull StorageReference ref,
      @NonNull File file) {
    return Single.create(new GetFileToFileOnSubscribe(ref, file));
  }

  /**
   * @see StorageReference#getMetadata()
   */
  public static Single<StorageMetadata> getMetadata(@NonNull StorageReference ref) {
    return Single.create(new GetMetadataOnSubscribe(ref));
  }

  /**
   * @see StorageReference#getStream()
   */
  public static Single<StreamDownloadTask.TaskSnapshot> getStream(@NonNull StorageReference ref) {
    return Single.create(new GetStreamOnSubscribe(ref));
  }

  /**
   * @see StorageReference#getStream(StreamDownloadTask.StreamProcessor)
   */
  public static Single<StreamDownloadTask.TaskSnapshot> getStream(@NonNull StorageReference ref,
      StreamDownloadTask.StreamProcessor streamProcessor) {
    return Single.create(new GetStreamWithProcessorOnSubscribe(ref, streamProcessor));
  }

  /**
   * @see StorageReference#getDownloadUrl()
   */
  public static Single<Uri> getDownloadUrl(@NonNull StorageReference ref) {
    return Single.create(new GetDownloadUriOnSubscribe(ref));
  }

  /**
   * @see StorageReference#putBytes(byte[])
   */
  public static Single<UploadTask.TaskSnapshot> putBytes(@NonNull StorageReference ref,
      byte[] bytes) {
    return Single.create(new PutBytesOnSubscribe(ref, bytes));
  }

  /**
   * @see StorageReference#putBytes(byte[], StorageMetadata)
   */
  public static Single<UploadTask.TaskSnapshot> putBytes(@NonNull StorageReference ref,
      byte[] bytes, StorageMetadata storageMetadata) {
    return Single.create(new PutBytesWithMetadataOnSubscribe(ref, bytes, storageMetadata));
  }

  /**
   * @see StorageReference#putFile(Uri)
   */
  public static Single<UploadTask.TaskSnapshot> putFile(@NonNull StorageReference ref,
      @NonNull Uri uri) {
    return Single.create(new PutFileOnSubscribe(ref, uri));
  }

  /**
   * @see StorageReference#putFile(Uri, StorageMetadata)
   */
  public static Single<UploadTask.TaskSnapshot> putFile(@NonNull StorageReference ref,
      @NonNull Uri uri, @NonNull StorageMetadata storageMetadata) {
    return Single.create(new PutFileWithMetadataOnSubscribe(ref, uri, storageMetadata));
  }

  /**
   * @see StorageReference#putFile(Uri, StorageMetadata, Uri)
   */
  public static Single<UploadTask.TaskSnapshot> putFile(@NonNull StorageReference ref,
      @NonNull Uri uri, @Nullable StorageMetadata storageMetadata,
      @Nullable Uri existingUploadUri) {
    return Single.create(new PutFileWithMetadataAndExistingUriOnSubscribe(ref, uri, storageMetadata,
        existingUploadUri));
  }

  /**
   * @see StorageReference#putStream(InputStream)
   */
  public static Single<UploadTask.TaskSnapshot> putStream(@NonNull StorageReference ref,
      @NonNull InputStream inputStream) {
    return Single.create(new PutStreamOnSubscribe(ref, inputStream));
  }

  /**
   * @see StorageReference#putStream(InputStream, StorageMetadata)
   */
  public static Single<UploadTask.TaskSnapshot> putStream(@NonNull StorageReference ref,
      @NonNull InputStream inputStream, @NonNull StorageMetadata storageMetadata) {
    return Single.create(new PutStreamWithMetadataOnSubscribe(ref, inputStream, storageMetadata));
  }

  /**
   * @see StorageReference#updateMetadata(StorageMetadata)
   */
  public static Single<StorageMetadata> updateMetadata(@NonNull StorageReference ref,
      @NonNull StorageMetadata storageMetadata) {
    return Single.create(new UpdateMetadataOnSubscribe(ref, storageMetadata));
  }
}

File: firebase-storage-kotlin/src/main/kotlin/com/github/b3er/rxfirebase/storage/RxFirebaseStorage.kt
@file:Suppress("NOTHING_TO_INLINE")

package com.github.b3er.rxfirebase.storage

import android.net.Uri
import com.google.firebase.storage.FileDownloadTask
import com.google.firebase.storage.StorageMetadata
import com.google.firebase.storage.StorageReference
import com.google.firebase.storage.StreamDownloadTask
import com.google.firebase.storage.StreamDownloadTask.StreamProcessor
import com.google.firebase.storage.UploadTask
import io.reactivex.Completable
import io.reactivex.Single
import java.io.File
import java.io.InputStream

inline fun StorageReference.rxGetBytes(maxDownloadSizeBytes: Long)
    : Single<ByteArray>
    = RxFirebaseStorage.getBytes(this, maxDownloadSizeBytes)

inline fun StorageReference.rxDelete()
    : Completable
    = RxFirebaseStorage.delete(this)

inline fun StorageReference.rxGetFile(uri: Uri)
    : Single<FileDownloadTask.TaskSnapshot>
    = RxFirebaseStorage.getFile(this, uri)

inline fun StorageReference.rxGetFile(file: File)
    : Single<FileDownloadTask.TaskSnapshot>
    = RxFirebaseStorage.getFile(this, file)

inline fun StorageReference.rxGetMetadata()
    : Single<StorageMetadata>
    = RxFirebaseStorage.getMetadata(this)

inline fun StorageReference.rxGetStream()
    : Single<StreamDownloadTask.TaskSnapshot>
    = RxFirebaseStorage.getStream(this)

inline fun StorageReference.rxGetStream(streamProcessor: StreamProcessor)
    : Single<StreamDownloadTask.TaskSnapshot>
    = RxFirebaseStorage.getStream(this, streamProcessor)

inline fun StorageReference.rxGetDownloadUrl()
    : Single<Uri>
    = RxFirebaseStorage.getDownloadUrl(this)

inline fun StorageReference.rxPutBytes(bytes: ByteArray)
    : Single<UploadTask.TaskSnapshot>
    = RxFirebaseStorage.putBytes(this, bytes)

inline fun StorageReference.rxPutBytes(bytes: ByteArray, storageMetadata: StorageMetadata)
    : Single<UploadTask.TaskSnapshot>
    = RxFirebaseStorage.putBytes(this, bytes, storageMetadata)

inline fun StorageReference.rxPutFile(uri: Uri)
    : Single<UploadTask.TaskSnapshot>
    = RxFirebaseStorage.putFile(this, uri)

inline fun StorageReference.rxPutFile(uri: Uri, storageMetadata: StorageMetadata)
    : Single<UploadTask.TaskSnapshot>
    = RxFirebaseStorage.putFile(this, uri, storageMetadata)

inline fun StorageReference.rxPutFile(uri: Uri, storageMetadata: StorageMetadata,
    existingUploadUri: Uri)
    : Single<UploadTask.TaskSnapshot>
    = RxFirebaseStorage.putFile(this, uri, storageMetadata, existingUploadUri)

inline fun StorageReference.rxPutStream(inputStream: InputStream)
    : Single<UploadTask.TaskSnapshot>
    = RxFirebaseStorage.putStream(this, inputStream)

inline fun StorageReference.rxPutStream(inputStream: InputStream, storageMetadata: StorageMetadata)
    : Single<UploadTask.TaskSnapshot>
    = RxFirebaseStorage.putStream(this, inputStream, storageMetadata)

inline fun StorageReference.rxUpdateMetadata(storageMetadata: StorageMetadata)
    : Single<StorageMetadata>
    = RxFirebaseStorage.updateMetadata(this, storageMetadata)